File: go/circuitrelay/transport.go
// SPDX-License-Identifier: MIT

package circuitrelay

import (
	"sync"
	"time"

	"github.com/google/uuid"
)

// RelayLink is the one-hop link a Transport uses to exchange raw relay frames with
// directly-reachable nodes — the seam between circuit-relay-v2 (transport-agnostic)
// and whatever real transport carries a frame one hop (BLE, Wi-Fi Direct, WebRTC,
// the HTTP relay, or an in-process link in tests). Mirrors the C# IRelayLink.
type RelayLink interface {
	// SendFrame sends a raw relay frame to a node reachable in one hop. Returns true
	// if the frame was handed to that node's link.
	SendFrame(node string, frame []byte) bool
	// CanReach reports whether this node currently has a direct one-hop link to node.
	CanReach(node string) bool
	// OnFrame registers the handler invoked when a raw frame arrives from a
	// directly-reachable node (sender node UHID, frame bytes).
	OnFrame(handler func(from string, frame []byte))
}

// Options tunes a Transport (mirrors C# CircuitRelayOptions).
type Options struct {
	ReservationTTL             time.Duration
	MaxReservations            int
	MaxBridges                 int
	BridgeDataLimitBytes       int64
	BridgeDurationLimitSeconds int32
	ConnectTimeout             time.Duration
	ReserveTimeout             time.Duration
	ActAsRelay                 bool
}

// DefaultOptions returns the same defaults as the C# reference.
func DefaultOptions() Options {
	return Options{
		ReservationTTL:  30 * time.Minute,
		MaxReservations: 128,
		MaxBridges:      128,
		ConnectTimeout:  10 * time.Second,
		ReserveTimeout:  10 * time.Second,
		ActAsRelay:      true,
	}
}

type relayBridge struct {
	a, b       string
	dataBudget int64
	deadline   time.Time // zero => no duration limit
	dataUsed   int64
	open       bool
}

type activeBridge struct {
	connID uuid.UUID
	relay  string
}

// Transport is the native circuit-relay-v2 engine: any node can act as target
// (Reserve), client (Send over a known relay route), and relay (grant reservations,
// bridge CONNECT→STOP, forward DATA under a budget). Faithful port of the C#
// CircuitRelayTransportService.
type Transport struct {
	localUhid string
	link      RelayLink
	opts      Options
	now       func() time.Time

	mu                  sync.Mutex
	reservations        map[string]time.Time      // relay role: client UHID -> expiry
	bridges             map[uuid.UUID]*relayBridge // relay role
	routes              map[string]string          // client: dest -> relay
	peerBridges         map[string]activeBridge    // endpoint: peer -> bridge
	pendingConnects     map[uuid.UUID]chan Status
	pendingReservations map[string]chan Status

	onData func(sender string, data []byte)
}

// NewTransport wires a Transport onto a link. now may be nil (defaults to time.Now).
func NewTransport(localUhid string, link RelayLink, opts Options, now func() time.Time) *Transport {
	if now == nil {
		now = time.Now
	}
	t := &Transport{
		localUhid:           localUhid,
		link:                link,
		opts:                opts,
		now:                 now,
		reservations:        map[string]time.Time{},
		bridges:             map[uuid.UUID]*relayBridge{},
		routes:              map[string]string{},
		peerBridges:         map[string]activeBridge{},
		pendingConnects:     map[uuid.UUID]chan Status{},
		pendingReservations: map[string]chan Status{},
	}
	link.OnFrame(t.onFrame)
	return t
}

// SetOnData registers the callback invoked when tunnelled data is delivered to this
// node as an endpoint (sender UHID, payload).
func (t *Transport) SetOnData(cb func(sender string, data []byte)) { t.onData = cb }

// SetRoute records that dest is reachable via relay (in production, from the
// directory / reservation gossip; tests set it directly).
func (t *Transport) SetRoute(dest, relay string) {
	t.mu.Lock()
	t.routes[dest] = relay
	t.mu.Unlock()
}

// ActiveBridgeCount / ActiveReservationCount are diagnostics for tests.
func (t *Transport) ActiveBridgeCount() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return len(t.bridges)
}
func (t *Transport) ActiveReservationCount() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return len(t.reservations)
}

// IsConnected reports whether a relay bridge to peer has been established.
func (t *Transport) IsConnected(peer string) bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	_, ok := t.peerBridges[peer]
	return ok
}

// Reserve reserves capacity on relay so peers can reach this node through it.
func (t *Transport) Reserve(relay string) bool {
	if !t.link.CanReach(relay) {
		return false
	}
	ch := make(chan Status, 1)
	t.mu.Lock()
	t.pendingReservations[relay] = ch
	t.mu.Unlock()
	defer func() {
		t.mu.Lock()
		delete(t.pendingReservations, relay)
		t.mu.Unlock()
	}()

	f := &RelayFrame{Type: MsgReserve, SourceUhid: t.localUhid, RelayUhid: relay}
	b, err := Serialize(f)
	if err != nil {
		return false
	}
	t.link.SendFrame(relay, b)
	return t.await(ch, t.opts.ReserveTimeout) == StatusOk
}

// Send delivers data to peer, establishing a relay bridge first if needed.
func (t *Transport) Send(peer string, data []byte) bool {
	t.mu.Lock()
	ab, ok := t.peerBridges[peer]
	t.mu.Unlock()
	if ok {
		return t.sendData(ab, peer, data)
	}

	t.mu.Lock()
	relay, hasRoute := t.routes[peer]
	t.mu.Unlock()
	if !hasRoute || !t.link.CanReach(relay) {
		return false
	}
	if t.connect(peer, relay) != StatusOk {
		return false
	}
	t.mu.Lock()
	ab, ok = t.peerBridges[peer]
	t.mu.Unlock()
	return ok && t.sendData(ab, peer, data)
}

func (t *Transport) connect(dest, relay string) Status {
	connID := uuid.New()
	ch := make(chan Status, 1)
	t.mu.Lock()
	t.pendingConnects[connID] = ch
	t.mu.Unlock()
	defer func() {
		t.mu.Lock()
		delete(t.pendingConnects, connID)
		t.mu.Unlock()
	}()

	f := &RelayFrame{
		Type:            MsgConnect,
		SourceUhid:      t.localUhid,
		DestinationUhid: dest,
		RelayUhid:       relay,
		ConnectionID:    connID.String(),
	}
	b, err := Serialize(f)
	if err != nil {
		return StatusConnectionFailed
	}
	if !t.link.SendFrame(relay, b) {
		return StatusConnectionFailed
	}
	return t.await(ch, t.opts.ConnectTimeout)
}

func (t *Transport) await(ch chan Status, timeout time.Duration) Status {
	select {
	case s := <-ch:
		return s
	case <-time.After(timeout):
		return StatusConnectionFailed
	}
}

func (t *Transport) sendData(ab activeBridge, peer string, data []byte) bool {
	f := &RelayFrame{
		Type:            MsgData,
		SourceUhid:      t.localUhid,
		DestinationUhid: peer,
		RelayUhid:       ab.relay,
		ConnectionID:    ab.connID.String(),
		Payload:         data,
	}
	b, err := Serialize(f)
	if err != nil {
		return false
	}
	return t.link.SendFrame(ab.relay, b)
}

// ── inbound dispatch ────────────────────────────────────────────────────────

func (t *Transport) onFrame(from string, frame []byte) {
	f, err := Deserialize(frame)
	if err != nil {
		return // drop malformed
	}
	switch f.Type {
	case MsgReserve:
		t.handleReserve(from, f)
	case MsgReserveResponse:
		t.handleReserveResponse(from, f)
	case MsgConnect:
		t.handleConnect(from, f)
	case MsgStop:
		t.handleStop(from, f)
	case MsgStopResponse:
		t.handleStopResponse(from, f)
	case MsgConnectResponse:
		t.handleConnectResponse(from, f)
	case MsgData:
		t.handleData(from, f)
	}
}

// Relay: grant/refuse a reservation.
func (t *Transport) handleReserve(from string, f *RelayFrame) {
	t.mu.Lock()
	if !t.opts.ActAsRelay || len(t.reservations) >= t.opts.MaxReservations {
		t.mu.Unlock()
		t.send(from, &RelayFrame{Type: MsgReserveResponse, SourceUhid: f.SourceUhid, RelayUhid: t.localUhid, Status: StatusReservationRefused})
		return
	}
	expiry := t.now().Add(t.opts.ReservationTTL)
	t.reservations[f.SourceUhid] = expiry
	t.mu.Unlock()
	t.send(from, &RelayFrame{
		Type: MsgReserveResponse, SourceUhid: f.SourceUhid, RelayUhid: t.localUhid,
		Status: StatusOk, ReservationExpiresAtMs: expiry.UnixMilli(),
	})
}

// Client: reservation confirmed/denied.
func (t *Transport) handleReserveResponse(from string, f *RelayFrame) {
	t.mu.Lock()
	ch, ok := t.pendingReservations[from]
	t.mu.Unlock()
	if ok {
		trySend(ch, f.Status)
	}
}

// Relay: A wants B. Validate B's reservation + reachability, open a STOP to B.
func (t *Transport) handleConnect(from string, f *RelayFrame) {
	a, b := f.SourceUhid, f.DestinationUhid
	connID, err := uuid.Parse(f.ConnectionID)
	if err != nil {
		return
	}
	if !t.opts.ActAsRelay {
		t.replyConnect(a, f, StatusConnectionFailed)
		return
	}
	t.mu.Lock()
	exp, has := t.reservations[b]
	if !has || !t.now().Before(exp) {
		delete(t.reservations, b)
		t.mu.Unlock()
		t.replyConnect(a, f, StatusNoReservation)
		return
	}
	if !t.link.CanReach(b) {
		t.mu.Unlock()
		t.replyConnect(a, f, StatusConnectionFailed)
		return
	}
	if len(t.bridges) >= t.opts.MaxBridges {
		t.mu.Unlock()
		t.replyConnect(a, f, StatusResourceLimitExceeded)
		return
	}
	var deadline time.Time
	if t.opts.BridgeDurationLimitSeconds > 0 {
		deadline = t.now().Add(time.Duration(t.opts.BridgeDurationLimitSeconds) * time.Second)
	}
	t.bridges[connID] = &relayBridge{a: a, b: b, dataBudget: t.opts.BridgeDataLimitBytes, deadline: deadline}
	t.mu.Unlock()

	t.send(b, &RelayFrame{
		Type: MsgStop, SourceUhid: a, DestinationUhid: b, RelayUhid: t.localUhid,
		ConnectionID: f.ConnectionID, LimitDataBytes: t.opts.BridgeDataLimitBytes,
		LimitDurationSeconds: t.opts.BridgeDurationLimitSeconds,
	})
}

// Target: relay says A wants us. Accept and record a return route to A.
func (t *Transport) handleStop(from string, f *RelayFrame) {
	connID, err := uuid.Parse(f.ConnectionID)
	if err != nil {
		return
	}
	t.mu.Lock()
	t.peerBridges[f.SourceUhid] = activeBridge{connID: connID, relay: from}
	t.mu.Unlock()
	t.send(from, &RelayFrame{
		Type: MsgStopResponse, SourceUhid: f.SourceUhid, DestinationUhid: t.localUhid,
		RelayUhid: from, ConnectionID: f.ConnectionID, Status: StatusOk,
	})
}

// Relay: target accepted/refused. Finalise the bridge and answer the client.
func (t *Transport) handleStopResponse(from string, f *RelayFrame) {
	connID, err := uuid.Parse(f.ConnectionID)
	if err != nil {
		return
	}
	t.mu.Lock()
	br, ok := t.bridges[connID]
	if !ok {
		t.mu.Unlock()
		return
	}
	if f.Status != StatusOk {
		aUhid := br.a
		delete(t.bridges, connID)
		t.mu.Unlock()
		t.replyConnect(aUhid, f, StatusConnectionFailed)
		return
	}
	br.open = true
	aUhid, bUhid, budget := br.a, br.b, br.dataBudget
	t.mu.Unlock()

	t.send(aUhid, &RelayFrame{
		Type: MsgConnectResponse, SourceUhid: aUhid, DestinationUhid: bUhid, RelayUhid: t.localUhid,
		ConnectionID: f.ConnectionID, Status: StatusOk, LimitDataBytes: budget,
	})
}

// Client: bridge established/refused.
func (t *Transport) handleConnectResponse(from string, f *RelayFrame) {
	connID, err := uuid.Parse(f.ConnectionID)
	if err != nil {
		return
	}
	if f.Status == StatusOk {
		t.mu.Lock()
		t.peerBridges[f.DestinationUhid] = activeBridge{connID: connID, relay: from}
		t.mu.Unlock()
	}
	t.mu.Lock()
	ch, ok := t.pendingConnects[connID]
	t.mu.Unlock()
	if ok {
		trySend(ch, f.Status)
	}
}

// Data: endpoint delivery, or relay forward (under budget).
func (t *Transport) handleData(from string, f *RelayFrame) {
	if f.DestinationUhid == t.localUhid {
		if t.onData != nil {
			t.onData(f.SourceUhid, f.Payload)
		}
		return
	}
	connID, err := uuid.Parse(f.ConnectionID)
	if err != nil {
		return
	}
	t.mu.Lock()
	br, ok := t.bridges[connID]
	if !ok || !br.open || (from != br.a && from != br.b) {
		t.mu.Unlock()
		return
	}
	if !br.deadline.IsZero() && !t.now().Before(br.deadline) {
		delete(t.bridges, connID)
		t.mu.Unlock()
		return
	}
	br.dataUsed += int64(len(f.Payload))
	over := br.dataBudget > 0 && br.dataUsed > br.dataBudget
	if over {
		delete(t.bridges, connID)
		t.mu.Unlock()
		return
	}
	t.mu.Unlock()

	if b, err := Serialize(f); err == nil {
		t.link.SendFrame(f.DestinationUhid, b) // forward unchanged to the other endpoint (its dst)
	}
}

// ── helpers ─────────────────────────────────────────────────────────────────

func (t *Transport) send(to string, f *RelayFrame) {
	if b, err := Serialize(f); err == nil {
		t.link.SendFrame(to, b)
	}
}

func (t *Transport) replyConnect(client string, connect *RelayFrame, status Status) {
	t.send(client, &RelayFrame{
		Type: MsgConnectResponse, SourceUhid: connect.SourceUhid, DestinationUhid: connect.DestinationUhid,
		RelayUhid: t.localUhid, ConnectionID: connect.ConnectionID, Status: status,
	})
}

func trySend(ch chan Status, s Status) {
	select {
	case ch <- s:
	default:
	}
}

File: typescript/src/circuitrelay/Transport.ts
/**
 * Native circuit-relay-v2 ENGINE — decentralised, no-libp2p any-node relaying on
 * the Aether mesh. Any node can act as target (Reserve capacity on a relay),
 * client (Send to a peer over a known relay route, performing the CONNECT
 * handshake then tunnelling DATA), and relay (grant reservations, bridge
 * CONNECT->STOP, and forward DATA between the two legs under a data/duration
 * budget) — all in this one class, a node can be any/all at once.
 *
 * Built on the fixture-locked {@link RelayFrame} wire format. One hop of a frame
 * is carried by the injected {@link RelayLink}. Faithful port of the C#
 * (AetherNet.Transport.CircuitRelay.CircuitRelayTransportService) and Go
 * (go/circuitrelay) reference engines — same roles, same validation, same
 * behavioural guarantees.
 *
 * SPDX-License-Identifier: MIT
 */

import {
  RelayFrame,
  RelayMessageType,
  RelayStatus,
  serializeRelayFrame,
  deserializeRelayFrame,
} from "./RelayFrame.js";

/**
 * The underlying one-hop link a {@link Transport} uses to exchange raw relay
 * frames with directly-reachable nodes — the seam between circuit-relay-v2
 * (transport-agnostic) and whatever real transport carries a frame one hop
 * (BLE, Wi-Fi Direct, WebRTC, the HTTP relay, or an in-process link in tests).
 * Mirrors the C# IRelayLink / Go RelayLink.
 */
export interface RelayLink {
  /**
   * Sends a raw relay frame to a node reachable in one hop. Returns true if the
   * frame was handed to that node's link.
   */
  sendFrame(node: string, frame: Uint8Array): boolean;
  /** Whether this node currently has a direct one-hop link to `node`. */
  canReach(node: string): boolean;
  /**
   * Registers the handler invoked when a raw frame arrives from a
   * directly-reachable node (sender node UHID, frame bytes).
   */
  onFrame(handler: (from: string, frame: Uint8Array) => void): void;
}

/** Tuning + policy for a {@link Transport} (mirrors C# CircuitRelayOptions / Go Options). */
export interface RelayOptions {
  /** How long a granted reservation remains valid, in milliseconds. */
  reservationTtlMs: number;
  /** Maximum concurrent reservations this node will hold as a relay. */
  maxReservations: number;
  /** Maximum concurrent bridges this node will service as a relay. */
  maxBridges: number;
  /** Per-bridge data budget in bytes granted by this relay. 0 = unlimited. */
  bridgeDataLimitBytes: number;
  /** Per-bridge duration budget in seconds granted by this relay. 0 = unlimited. */
  bridgeDurationLimitSeconds: number;
  /** How long a client waits for a CONNECT to be confirmed before giving up, in ms. */
  connectTimeoutMs: number;
  /** How long a client waits for a RESERVE to be confirmed before giving up, in ms. */
  reserveTimeoutMs: number;
  /** Whether this node grants reservations and bridges traffic for others. */
  actAsRelay: boolean;
}

/** The same defaults as the C# / Go references. */
export function defaultRelayOptions(): RelayOptions {
  return {
    reservationTtlMs: 30 * 60 * 1000, // 30 minutes
    maxReservations: 128,
    maxBridges: 128,
    bridgeDataLimitBytes: 0,
    bridgeDurationLimitSeconds: 0,
    connectTimeoutMs: 10 * 1000,
    reserveTimeoutMs: 10 * 1000,
    actAsRelay: true,
  };
}

/** A bridge this node is relaying (relay role). */
interface RelayBridge {
  a: string;
  b: string;
  dataBudget: number;
  /** Epoch ms deadline; Number.POSITIVE_INFINITY => no duration limit. */
  deadline: number;
  dataUsed: number;
  open: boolean;
}

/** An established bridge from this node's endpoint view: which connection, via which relay. */
interface ActiveBridge {
  connId: string;
  relay: string;
}

/** A CONNECT/RESERVE waiter: a resolver plus the timer to cancel on resolution. */
interface Pending {
  resolve: (status: RelayStatus) => void;
  timer: ReturnType<typeof setTimeout>;
}

/**
 * The native circuit-relay-v2 engine. Wire it onto a {@link RelayLink}; register
 * an onData callback to receive tunnelled data delivered to this node as an
 * endpoint.
 */
export class Transport {
  private readonly localUhid: string;
  private readonly link: RelayLink;
  private readonly opts: RelayOptions;
  private readonly now: () => number;

  // Relay role
  private readonly reservations = new Map<string, number>(); // client UHID -> expiry (epoch ms)
  private readonly bridges = new Map<string, RelayBridge>(); // connId -> bridge

  // Client / target role
  private readonly routes = new Map<string, string>(); // dest -> relay
  private readonly peerBridges = new Map<string, ActiveBridge>(); // peer -> bridge
  private readonly pendingConnects = new Map<string, Pending>(); // connId -> waiter
  private readonly pendingReservations = new Map<string, Pending>(); // relay -> waiter

  private onDataCb: ((sender: string, data: Uint8Array) => void) | null = null;

  /**
   * @param localUhid This node's UHID.
   * @param link One-hop link to directly-reachable nodes.
   * @param opts Policy/tuning (defaults to {@link defaultRelayOptions}).
   * @param now Clock returning epoch ms (injectable for deterministic
   *   reservation-expiry tests; defaults to Date.now).
   */
  constructor(
    localUhid: string,
    link: RelayLink,
    opts: RelayOptions = defaultRelayOptions(),
    now: () => number = () => Date.now(),
  ) {
    this.localUhid = localUhid;
    this.link = link;
    this.opts = opts;
    this.now = now;
    this.link.onFrame((from, frame) => this.onFrame(from, frame));
  }

  /**
   * Registers the callback invoked when tunnelled data is delivered to this node
   * as an endpoint (sender UHID, payload).
   */
  setOnData(cb: (sender: string, data: Uint8Array) => void): void {
    this.onDataCb = cb;
  }

  /**
   * Records that `dest` is reachable via `relay` (in production, from the
   * directory / reservation gossip; tests set it directly).
   */
  setRoute(dest: string, relay: string): void {
    this.routes.set(dest, relay);
  }

  /** Number of bridges this node is currently servicing as a relay (diagnostics/tests). */
  activeBridgeCount(): number {
    return this.bridges.size;
  }

  /** Number of reservations this node is currently holding as a relay (diagnostics/tests). */
  activeReservationCount(): number {
    return this.reservations.size;
  }

  /** True once a relay bridge to `peer` has been established. */
  isConnected(peer: string): boolean {
    return this.peerBridges.has(peer);
  }

  /**
   * Reserves capacity on `relay` so peers can reach this node through it.
   * Resolves true once the relay confirms the reservation.
   */
  async reserve(relay: string): Promise<boolean> {
    if (!this.link.canReach(relay)) return false;

    const wait = this.registerPending(this.pendingReservations, relay, this.opts.reserveTimeoutMs);
    const frame: RelayFrame = this.newFrame(RelayMessageType.Reserve, {
      sourceUhid: this.localUhid,
      relayUhid: relay,
    });
    this.link.sendFrame(relay, serializeRelayFrame(frame));

    try {
      return (await wait) === RelayStatus.Ok;
    } finally {
      this.pendingReservations.delete(relay);
    }
  }

  /**
   * Delivers `data` to `peer`, establishing a relay bridge first if needed.
   * Resolves true if the frame was handed to the relay.
   */
  async send(peer: string, data: Uint8Array): Promise<boolean> {
    const existing = this.peerBridges.get(peer);
    if (existing) return this.sendData(existing, peer, data);

    // No bridge yet — establish one through the known relay for this peer.
    const relay = this.routes.get(peer);
    if (relay === undefined || !this.link.canReach(relay)) return false;

    if ((await this.connect(peer, relay)) !== RelayStatus.Ok) return false;

    const b = this.peerBridges.get(peer);
    return b !== undefined && this.sendData(b, peer, data);
  }

  // ── Client handshake ────────────────────────────────────────────────────────

  private async connect(dest: string, relay: string): Promise<RelayStatus> {
    const connId = crypto.randomUUID();
    const wait = this.registerPending(this.pendingConnects, connId, this.opts.connectTimeoutMs);
    const frame: RelayFrame = this.newFrame(RelayMessageType.Connect, {
      sourceUhid: this.localUhid,
      destinationUhid: dest,
      relayUhid: relay,
      connectionId: connId,
    });

    try {
      if (!this.link.sendFrame(relay, serializeRelayFrame(frame))) return RelayStatus.ConnectionFailed;
      return await wait;
    } finally {
      this.pendingConnects.delete(connId);
    }
  }

  private sendData(bridge: ActiveBridge, peer: string, data: Uint8Array): boolean {
    const frame: RelayFrame = this.newFrame(RelayMessageType.Data, {
      sourceUhid: this.localUhid,
      destinationUhid: peer,
      relayUhid: bridge.relay,
      connectionId: bridge.connId,
      payload: data,
    });
    return this.link.sendFrame(bridge.relay, serializeRelayFrame(frame));
  }

  // ── Inbound frame dispatch ────────────────────────────────────────────────────

  private onFrame(from: string, frame: Uint8Array): void {
    let f: RelayFrame;
    try {
      f = deserializeRelayFrame(frame);
    } catch {
      return; // drop malformed
    }
    switch (f.type) {
      case RelayMessageType.Reserve:
        this.handleReserve(from, f);
        break;
      case RelayMessageType.ReserveResponse:
        this.handleReserveResponse(from, f);
        break;
      case RelayMessageType.Connect:
        this.handleConnect(from, f);
        break;
      case RelayMessageType.Stop:
        this.handleStop(from, f);
        break;
      case RelayMessageType.StopResponse:
        this.handleStopResponse(from, f);
        break;
      case RelayMessageType.ConnectResponse:
        this.handleConnectResponse(from, f);
        break;
      case RelayMessageType.Data:
        this.handleData(from, f);
        break;
    }
  }

  // Relay: grant/refuse a reservation.
  private handleReserve(from: string, f: RelayFrame): void {
    if (!this.opts.actAsRelay || this.reservations.size >= this.opts.maxReservations) {
      this.send2(from, this.newFrame(RelayMessageType.ReserveResponse, {
        sourceUhid: f.sourceUhid,
        relayUhid: this.localUhid,
        status: RelayStatus.ReservationRefused,
      }));
      return;
    }
    const expiry = this.now() + this.opts.reservationTtlMs;
    this.reservations.set(f.sourceUhid, expiry);
    this.send2(from, this.newFrame(RelayMessageType.ReserveResponse, {
      sourceUhid: f.sourceUhid,
      relayUhid: this.localUhid,
      status: RelayStatus.Ok,
      reservationExpiresAtMs: expiry,
    }));
  }

  // Client: reservation confirmed/denied.
  private handleReserveResponse(from: string, f: RelayFrame): void {
    this.resolvePending(this.pendingReservations, from, f.status);
  }

  // Relay: A wants to reach B. Validate B's reservation + reachability, then open a STOP to B.
  private handleConnect(from: string, f: RelayFrame): void {
    const a = f.sourceUhid;
    const b = f.destinationUhid;

    if (!this.opts.actAsRelay) {
      this.replyConnect(a, f, RelayStatus.ConnectionFailed);
      return;
    }
    const exp = this.reservations.get(b);
    if (exp === undefined || this.now() >= exp) {
      this.reservations.delete(b);
      this.replyConnect(a, f, RelayStatus.NoReservation);
      return;
    }
    if (!this.link.canReach(b)) {
      this.replyConnect(a, f, RelayStatus.ConnectionFailed);
      return;
    }
    if (this.bridges.size >= this.opts.maxBridges) {
      this.replyConnect(a, f, RelayStatus.ResourceLimitExceeded);
      return;
    }

    const deadline =
      this.opts.bridgeDurationLimitSeconds > 0
        ? this.now() + this.opts.bridgeDurationLimitSeconds * 1000
        : Number.POSITIVE_INFINITY;
    this.bridges.set(f.connectionId, {
      a,
      b,
      dataBudget: this.opts.bridgeDataLimitBytes,
      deadline,
      dataUsed: 0,
      open: false,
    });

    this.send2(b, this.newFrame(RelayMessageType.Stop, {
      sourceUhid: a,
      destinationUhid: b,
      relayUhid: this.localUhid,
      connectionId: f.connectionId,
      limitDataBytes: this.opts.bridgeDataLimitBytes,
      limitDurationSeconds: this.opts.bridgeDurationLimitSeconds,
    }));
  }

  // Target: relay says A wants to reach us. Accept and record a return route to A.
  private handleStop(from: string, f: RelayFrame): void {
    this.peerBridges.set(f.sourceUhid, { connId: f.connectionId, relay: from });
    this.send2(from, this.newFrame(RelayMessageType.StopResponse, {
      sourceUhid: f.sourceUhid,
      destinationUhid: this.localUhid,
      relayUhid: from,
      connectionId: f.connectionId,
      status: RelayStatus.Ok,
    }));
  }

  // Relay: target accepted/refused. Finalise the bridge and answer the client.
  private handleStopResponse(from: string, f: RelayFrame): void {
    const bridge = this.bridges.get(f.connectionId);
    if (bridge === undefined) return;

    if (f.status !== RelayStatus.Ok) {
      this.bridges.delete(f.connectionId);
      this.replyConnect(bridge.a, f, RelayStatus.ConnectionFailed);
      return;
    }

    bridge.open = true;
    this.send2(bridge.a, this.newFrame(RelayMessageType.ConnectResponse, {
      sourceUhid: bridge.a,
      destinationUhid: bridge.b,
      relayUhid: this.localUhid,
      connectionId: f.connectionId,
      status: RelayStatus.Ok,
      limitDataBytes: bridge.dataBudget,
    }));
  }

  // Client: bridge established/refused.
  private handleConnectResponse(from: string, f: RelayFrame): void {
    if (f.status === RelayStatus.Ok) {
      this.peerBridges.set(f.destinationUhid, { connId: f.connectionId, relay: from });
    }
    this.resolvePending(this.pendingConnects, f.connectionId, f.status);
  }

  // Data: either I'm an endpoint (deliver) or the relay (forward the other way, under budget).
  private handleData(from: string, f: RelayFrame): void {
    if (f.destinationUhid === this.localUhid) {
      this.onDataCb?.(f.sourceUhid, f.payload);
      return;
    }

    const bridge = this.bridges.get(f.connectionId);
    if (bridge === undefined || !bridge.open) return; // unknown / not-yet-open bridge — drop
    if (from !== bridge.a && from !== bridge.b) return; // frame not from a party to this bridge

    if (this.now() >= bridge.deadline) {
      this.bridges.delete(f.connectionId);
      return;
    }

    bridge.dataUsed += f.payload.length;
    if (bridge.dataBudget > 0 && bridge.dataUsed > bridge.dataBudget) {
      this.bridges.delete(f.connectionId);
      return;
    }

    // Forward the frame unchanged to the other endpoint (= its dst).
    this.link.sendFrame(f.destinationUhid, serializeRelayFrame(f));
  }

  // ── Pending-map helpers (Promise + timeout race) ──────────────────────────────

  private registerPending(
    map: Map<string, Pending>,
    key: string,
    timeoutMs: number,
  ): Promise<RelayStatus> {
    return new Promise<RelayStatus>((resolve) => {
      const timer = setTimeout(() => {
        map.delete(key);
        resolve(RelayStatus.ConnectionFailed); // timeout
      }, timeoutMs);
      map.set(key, { resolve, timer });
    });
  }

  private resolvePending(map: Map<string, Pending>, key: string, status: RelayStatus): void {
    const p = map.get(key);
    if (p === undefined) return;
    map.delete(key);
    clearTimeout(p.timer);
    p.resolve(status);
  }

  // ── Frame construction / send helpers ─────────────────────────────────────────

  private newFrame(type: RelayMessageType, fields: Partial<RelayFrame>): RelayFrame {
    return {
      type,
      status: fields.status ?? RelayStatus.Ok,
      sourceUhid: fields.sourceUhid ?? "",
      destinationUhid: fields.destinationUhid ?? "",
      relayUhid: fields.relayUhid ?? "",
      connectionId: fields.connectionId ?? "",
      reservationExpiresAtMs: fields.reservationExpiresAtMs ?? 0,
      limitDurationSeconds: fields.limitDurationSeconds ?? 0,
      limitDataBytes: fields.limitDataBytes ?? 0,
      payload: fields.payload ?? new Uint8Array(0),
    };
  }

  private send2(to: string, f: RelayFrame): void {
    this.link.sendFrame(to, serializeRelayFrame(f));
  }

  private replyConnect(client: string, connect: RelayFrame, status: RelayStatus): void {
    this.send2(client, this.newFrame(RelayMessageType.ConnectResponse, {
      sourceUhid: connect.sourceUhid,
      destinationUhid: connect.destinationUhid,
      relayUhid: this.localUhid,
      connectionId: connect.connectionId,
      status,
    }));
  }
}
